relayer/validate.go

```diff
@@ -0,0 +1,127 @@
+package relayer
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/athanorlabs/go-relayer/impls/gsnforwarder"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethclient"
+
+	"github.com/athanorlabs/atomic-swap/coins"
+	"github.com/athanorlabs/atomic-swap/common/types"
+	contracts "github.com/athanorlabs/atomic-swap/ethereum"
+	"github.com/athanorlabs/atomic-swap/net/message"
+)
+
+func validateClaimRequest(
+	ctx context.Context,
+	request *message.RelayClaimRequest,
+	ec *ethclient.Client,
+	ourSFContractAddr ethcommon.Address,
+) error {
+	err := validateClaimValues(ctx, request, ec, ourSFContractAddr)
+	if err != nil {
+		return err
+	}
+
+	return validateClaimSignature(ctx, ec, request)
+}
+
+// validateClaimValues validates the non-signature aspects of the claim request:
+//  1. the claim request swap factory contract is byte compatible with ours
+//  2. the forwarder in the claim request swap factory contract has an identical
+//     address with our forwarder
+//  3. the relayer fee is equal to or greater than the passed minFee
+//  4. the swap is for ETH and not an ERC20 token
+//  5. the swap value is strictly greater than the relayer fee
+//  6. TODO: Validate that the swap exists and is in a claimable state?
```

Comment:
this is a good idea, I'd add it in

Reply:
Yeah, I'd like to see what happens now when we send a request, if the swap is not in a claimable state. In theory, the ethereum client simulates the transaction using the data (e.g. block timestamp) from the last block received. If the transaction reverts during simulation, it won't be sent to the mempool and no gas is lost. If this is the case, then the additional validation is probably a low priority.

We have seen cases where the transaction failed after it went out to the mempool. It would happen because the mined block's timestamp was further into the future than the last block triggering a revert. We had to write extra code to extract out the revert message when that happens.